### PR TITLE
Issue #11214: Specified violation message in WriteTagCheck

### PR DESCRIPTION
**Pull request by Sloent**
Issue #11214 : Specified violation message in WriteTagCheck

### CODE REVIEW

**Comment by Sloent** [comment]
I know how to explain why the various Input files of WriteTag have errors.  But every time I add a character to //violation in javadoc, the test fails.  Even if I adjusted the value of the exception array in "src\main\java\com\puppycrawl\tools\checkstyle\checks\javadoc\WriteTagCheck.java" and the value of tagFormat accordingly, it didn't work Tested.

I think that modifying and stating a violation in javadoc is one more level of restriction than other checktests.  Can I open another line that does not contain a violation comment to explain why the error occurred?

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/javadoc/WriteTagCheckTest.java [hunk 1]
```diff
@@ -58,32 +58,33 @@ public void testDefaultSettings() throws Exception {
     @Test
     public void testTag() throws Exception {
         final String[] expected = {
-            "15: " + getCheckMessage(MSG_WRITE_TAG, "@author", "Daniel Grenner"),
+            "21: " + getCheckMessage(MSG_WRITE_TAG, "@author", "// violation 3 lines" +
+                    " below 'Must contain non-whitespace characters.'"),
```
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/javadoc/WriteTagCheckTest.java:62]: you do not need to change expected.
look at https://github.com/checkstyle/checkstyle/commit/9857df8546c56f48b68fc5e930caffcc70eafe01 and other example of same type of update to see what others are doing.

**Review by romani — CHANGES_REQUESTED** [review]
items:

**Comment by Sloent** [comment]
I restored the WriteTagCheck.java code to its original form,
@Test
    public void testRegularExError() throws Exception {
        final String[] expected = {
            "15: " + getCheckMessage(MSG_TAG_FORMAT, "@author", "ABC"),
        };
        verifyWithInlineConfigParser(getPath("InputWriteTagExpressionError.java"), expected);
    }
    
/**
 * Testing tag writing
 * @author Daniel Grenner  // violation （15 lines）
 * @incomplete This class needs more code...
 * @doubletag first text
 * @doubletag second text
 * @emptytag
 */
 
The top code becomes the bottom code:

 // violation 'The tagFormat requires ABC, not Daniel Grenner.' （12 lines）
/**
 * Testing tag writing
 * @author Daniel Grenner 
 * @incomplete This class needs more code...
 * @doubletag first text
 * @doubletag second text
 * @emptytag
 */
 
error：
expected: 12
but was : 15

**Comment by romani** [comment]
@Sloent , please never do development in foked master, always do branch
https://checkstyle.org/beginning_development.html#Starting_Development

in other case you will be very slow.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/writetag/InputWriteTagExpressionError.java [hunk 2]
```diff
@@ -12,7 +12,7 @@
 
 /**
  * Testing tag writing
- * @author Daniel Grenner  // violation
+ * @author Daniel Grenner  // violation 'The tagFormat requires ABC, not Daniel Grenner.'
```
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/writetag/InputWriteTagExpressionError.java:15]: move all trailing comments out of javadoc comments. use `// violation 7 lines above|below `

**Review by romani — CHANGES_REQUESTED** [review]
items:

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/writetag/InputWriteTagExpressionError.java (hunk 2 above)
Comment (Sloent) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/writetag/InputWriteTagExpressionError.java:15]: It would seem that whenever a comment contains the word "// violation", the line is considered to be an expected value, thus causing an error. But if I just wrote "//" plus any character other than "violation", there would be no error.
In this case, if I had used // violation 7 lines above, the expected value would have been reduced by 7 lines accordingly, thus causing an error.

Comment (romani) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/writetag/InputWriteTagExpressionError.java:15]: this Check validate comment content, do not change comment content by such technical markers.

Comment (Sloent) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/writetag/InputWriteTagExpressionError.java:15]: /**
 * Testing tag writing
 * @author Daniel Grenner  // violation
 
If I change the above code to the following code:
 
// violation 3 lines below 'The tagFormat requires ABC, not Daniel Grenner.'
 /**
 * Testing tag writing
 * @author Daniel Grenner 
 
The test will say:
 Actual and expected violations differ.
 expected to match: 15:(?:\d+:)?\s.*The tagFormat requires ABC, not Daniel Grenner..* 
 but was: 15: The tag @author must match: 'ABC'.
 
 Do you have any idea how to solve this problem?

Comment (romani) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/writetag/InputWriteTagExpressionError.java:15]: Keep debugging test execution

Comment (Sloent) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/writetag/InputWriteTagExpressionError.java:15]: final String[] expected = {
            "15: " + getCheckMessage(MSG_TAG_FORMAT, "@author", "ABC")
        };
        
According to the expected array in the test method, the violation comment can only match an actual error in Chinese, but not in English. What should I do?

// violation 3 lines below '标签 @author 必须匹配： 'ABC' 。'
/**
 * Testing tag writing
 * @author Daniel Grenner
 (success!)
--------------------------------------------------------------------------------------------------------------
 // violation 3 lines below '@author must match 'ABC'.'
/**
 * Testing tag writing
 * @author Daniel Grenner
(failure)
 
 Actual and expected violations differ.
expected to match: 15:(?:\d+:)?\s.*@author must match 'ABC'..*
but was          : 15: 标签 @author 必须匹配： 'ABC' 。

Comment (romani) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/writetag/InputWriteTagExpressionError.java:15]: Ok, we run into locale problem https://github.com/checkstyle/checkstyle/issues/12104

Please enforce English locale for test execution.
Sorry for inconvenience

Comment (Sloent) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/writetag/InputWriteTagExpressionError.java:15]: I see. Thank you

**Comment by Sloent** [comment]
I will have an algorithm contest on April 8th, and I have to modify the proposal and preparation contest in these days. It is expected that I will continue to repair the code on April 9th to pass the CI check.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/writetag/InputWriteTagExpressionError.java [hunk 3]
```diff
@@ -9,10 +9,10 @@
 */
 
 package com.puppycrawl.tools.checkstyle.checks.javadoc.writetag;
-
+// violation 3 lines below '标签 @author 必须匹配： 'ABC' 。'
```
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/writetag/InputWriteTagExpressionError.java:12]: In English locale please.

**Review by romani — CHANGES_REQUESTED** [review]
Items

**Comment by romani** [comment]
Fix CI problems
https://ci.appveyor.com/project/Checkstyle/checkstyle/builds/46794563/job/09yt7f44694ojwlt#L104

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/javadoc/writetag/InputWriteTagRecordsAndCompactCtors.java [hunk 4]
```diff
@@ -11,12 +11,12 @@
...
 
-// violation below , 'Type Javadoc comment is missing @incomplete tag'
+// violation 1 line below , 'Type Javadoc comment is missing @incomplete tag.*'
 public class InputWriteTagRecordsAndCompactCtors { // violation
 
-    // violation 2 lines below 'Type Javadoc tag @incomplete must match pattern '\\S''
+    // violation 2 lines below , 'Type Javadoc tag @incomplete must match pattern '\\S'.*'
     /**
-     * @incomplete
+     * @incomplete // violation
```
Comment (romani) [inline src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/javadoc/writetag/InputWriteTagRecordsAndCompactCtors.java:19]: Remove violation wording or change it to some text

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/writetag/InputWriteTagEnumsAndAnnotations.java [hunk 5]
```diff
@@ -9,7 +9,7 @@
 */
 
 package com.puppycrawl.tools.checkstyle.checks.javadoc.writetag;
-
+// violation 3, 7, 13, 17 lines below , 'Cannot be empty.*'
```
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/writetag/InputWriteTagEnumsAndAnnotations.java:12]: Bring it close to javadoc with violation

**Review by romani — CHANGES_REQUESTED** [review]
Items

**Comment by Sloent** [comment]
I'm solving the "local master branch is 28 commits worse than checkstyle's branch" problem.   So I enter the following command to update the local branch:
git fetch upstream
git merge upstream/master
git push origin master
As a result, I became a commit contributor to all 28 of them (thankfully without tampering with their code).   I am very, very sorry for the inconvenience brought to you.   May I ask how can I cancel my submission in other people's commit?

**Comment by romani** [comment]
Backup your changes on local . Remove local branch. https://checkstyle.org/beginning_development.html#Starting_Development
So all from scratch. Push in GitHub to same branch with force

**Comment by Sloent** [comment]
I solved problems temporarily

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/javadoc/WriteTagCheckTest.java [hunk 6]
```diff
@@ -169,6 +170,7 @@ public void testNoJavadocs() throws Exception {
 
     @Test
     public void testWriteTagRecordsAndCompactCtors() throws Exception {
+        Locale.setDefault(Locale.ENGLISH);
         final String[] expected = {
```
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/javadoc/WriteTagCheckTest.java:174]: Please revert this change.

Please remove this Check from suppression list as all others done in their PRs for such update.

**Review by romani — CHANGES_REQUESTED** [review]
Last:

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/javadoc/WriteTagCheckTest.java (hunk 6 above)
Comment (Sloent) [reply src/test/java/com/puppycrawl/tools/checkstyle/checks/javadoc/WriteTagCheckTest.java:174]: I delete this check from the ban list "com.puppycrawl.tools.checkstyle.checks.javadoc.WriteTagCheck",  and then Locale.setDefault(Locale.ENGLISH);  needs to be set separately in each test method,  otherwise the expected error message will be in my own language, and the English error message will not pass the test.

I tried to change my compilation output to English by adding '-Duser.language=en' to idea64.exe. vmoptions, idea. properties, java compiler in idea Settings and system user variable, but it didn't work.  It can be said that my problem is stuck in that I do not know how to modify the output information of IDEA compilation into English. What do you have in mind?

**Comment by Sloent** [comment]
After removing "com.puppycrawl.tools.checkstyle.checks.javadoc.WriteTagCheck", if a javadoc comment has the //violation identifier, the format requires that the right side of //violation must specify the cause of the error, otherwise it will fail the test.   But you prompted me earlier to move all trailing comments out of the javadoc comments. Using "// violation 7 lines above | below ".

So can I imitate #12953 and delete the //violation flag, and then add the violation cause above or below it? As shown below:
image

**Comment by romani** [comment]
>delete the //violation flag, and then add the violation cause above or below it?

Exactly what we need.

**Comment by romani** [comment]
Fyi, if we merge your changes sooner it will help us at https://github.com/checkstyle/checkstyle/pull/12819#discussion_r1173780936

**Comment by Sloent** [comment]
I used two pictures to ask two questions. After horizontal comparison and debugging, I still don't know how to solve it...

image

image

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/javadoc/writetag/InputWriteTagRecordsAndCompactCtors.java [hunk 7]
```diff
@@ -21,45 +21,45 @@ public class InputWriteTagRecordsAndCompactCtors { // violation
...
-    // violation below , 'Type Javadoc comment is missing @incomplete tag'
-    record MyRecord3(int x) { // violation
 
-        // violation 2 lines below , 'Javadoc tag .*'
+    record MyRecord3(int x) { // violation 'Type Javadoc comment is missing @incomplete tag.*'
+
+        // violation 2 lines below 'Failed to recognize 'record' introduced in Java 14.'
         /**
-         * @incomplete // violation
+         * @incomplete Failed to recognize 'record' introduced in Java 14.
```
Comment (romani) [inline src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/javadoc/writetag/InputWriteTagRecordsAndCompactCtors.java:48]: it sounds like a defect , right ?
if you think so, please created issue on this 

.... and you can fix it as your next issue, as you might be already so familiar with this Check, it should be easy win.

**Review by romani — APPROVED** [review]
ok to merge if CI pass.

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/javadoc/writetag/InputWriteTagRecordsAndCompactCtors.java (hunk 7 above)
Comment (romani) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/javadoc/writetag/InputWriteTagRecordsAndCompactCtors.java:48]: @Sloent , please do this

Comment (Sloent) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/javadoc/writetag/InputWriteTagRecordsAndCompactCtors.java:48]: First of all, thank you very much for your constructive suggestions during this process.  Moreover, this is also my first attempt to deal with your new question.  Please allow me to ask the following questions:

①Are you saying that you want me to make further changes so that java versions 14 and up can pass the test?

②Do issues have to be submitted simultaneously with PR?

③If I want to make sure I'm on the right track, or if there's anything I need to add, where should I discuss it?

Comment (romani) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/javadoc/writetag/InputWriteTagRecordsAndCompactCtors.java:48]: Please create new  issue in our issue tracker.
I will approve it, and you can send PR to fix it.
Or maybe send PR after you fix gsoc third issues.

Comment (romani) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/javadoc/writetag/InputWriteTagRecordsAndCompactCtors.java:48]: @Sloent , please create issue.

Comment (Sloent) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/javadoc/writetag/InputWriteTagRecordsAndCompactCtors.java:48]: Ok, I'll bring up the issue tonight

Comment (Sloent) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/javadoc/writetag/InputWriteTagRecordsAndCompactCtors.java:48]: I have created it, please approve

Comment (Sloent) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/javadoc/writetag/InputWriteTagRecordsAndCompactCtors.java:48]:
```
// violation 2 lines below 'Failed to recognize 'record' introduced in Java 14.'
    /**
     * @incomplete Failed to recognize 'record' introduced in Java 14.
     */
    record MyRecord1() {

    }
```
The reason for the error in this code is that the 'record' keyword is an illegal identifier in Java 11-13. Therefore, when the compiler encounters the 'record' keyword, it throws an 'illegal start of expression' compilation error.

In this case, the Java version that compiles 'record' must be version 14 or later. This way the following code will not report an error:

```
/**
 * @incomplete This class needs more code...
 */
 record MyRecord1() {
 
 }
```

Therefore, in order to fill in the gaps, I think it is OK to make the test cases that comply with the rule "' @incomplete 'must be followed by a non-empty description" legal.
